Gems/Atom/Feature/Common/Code/Source/Mesh/ModelReloader.cpp

```diff
@@ -0,0 +1,173 @@
+/*
+ * Copyright (c) Contributors to the Open 3D Engine Project.
+ * For complete copyright and license terms please see the LICENSE at the root of this distribution.
+ *
+ * SPDX-License-Identifier: Apache-2.0 OR MIT
+ *
+ */
+
+#include <Source/Mesh/ModelReloader.h>
+#include <Atom/RPI.Reflect/Model/ModelAsset.h>
+#include <Atom/RPI.Public/Model/Model.h>
+
+namespace AZ
+{
+    namespace Render
+    {
+        ModelReloader::ModelReloader(
+            Data::Asset<RPI::ModelAsset> modelAsset, RemoveModelFromReloaderSystemEvent::Handler& removeReloaderFromSystemHandler)
+        {
+            m_modelAsset.push_back(modelAsset);
+            m_pendingDependencies.reset();
+            removeReloaderFromSystemHandler.Connect(m_onRemoveReloaderFromSystem);
+
+            // Iterate over the model and track the assets that need to be reloaded
+            for (auto& modelLodAsset : modelAsset->GetLodAssets())
+            {
+                for (auto& mesh : modelLodAsset->GetMeshes())
+                {
+                    for (auto& streamBufferInfo : mesh.GetStreamBufferInfoList())
+                    {
+                        InsertMeshDependencyIfUnique(streamBufferInfo.m_bufferAssetView.GetBufferAsset());
+                        
+                    }
+                    InsertMeshDependencyIfUnique(mesh.GetIndexBufferAssetView().GetBufferAsset());
+                }
+                m_modelDependencies.push_back(modelLodAsset);
+            }
```

Comment:
While I have no doubt that this was tested thoroughly, this seems to only be tracking dependencies of the current model.  Can you please elaborate on how this accounts for the current model being replaced by one with a completely different layout, LODs, streams, and asset IDs?  For example, what happens if reloading alternates between sponza and a shader ball?

Could this possibly be reloading dependencies that might not exist on the reloaded asset?  For example, going from sponza with several dependencies to the shader ball with very few, this would attempt to reload everything from sponza unnecessarily.

I think these potential problems can be avoided and the code can be optimized by doing things in the reverse order.  This would also allow the same code to be used for loading and reloading, guaranteeing that all/only the necessary dependencies are loaded.

1.	Queue and connect the top level model asset.  Wait for completion.
2.	Queue and connect all the LOD assets from the new model.  Wait for completion.
3.	Queue and connect all the buffer assets from the new model.  Wait for completion.

This might also mean that you can keep the auto loading support in the individual asset types.  Calling QueueLoad for something that is already loaded with the current data shouldn’t do anything but trigger the notification.

Reply:
Yes, this can be reloading lods that no longer exist in the new asset, in which case you get an OnAssetError event. The ModelReloader::OnAssetError event handler will then mark that asset as complete.

While my approach works, I do think your suggestion makes sense, in which case instead of determining all the assets that need to be reloaded up front, I just don't check which assets need to be reloaded until the higher level asset is finished and I'm transitioning between hierarchies. I will try that, after which I should be able to have OnAssetReloadError actually display an error.

The testing details are in https://github.com/o3de/o3de-atomtest/pull/78

Reply:
I recall now why this order didn't work. Once the ModelAsset is reloaded, it will have references to the out of date ModelLodAsset. So even if you queue the reload of the ModelLodAssets at that time, which will have the correct number of lods and refer to the correct assetIds, for any lods that previously existed, it will be referring to an out of date ModelLodAsset. Once you queue the ModelLodAsset, it will get reloaded, but the ModelAsset will not end up with the updated references. Same thing happens when going from ModelLodAsset->BufferAsset level of the hierarchy.

We could add some plumbing to update the ModelAsset when that happens by moving some of that logic to the ModelAsset class, as you suggest handling it in the individual asset types. But we would still need some tracking to keep track of when all the dependencies are reloaded, so that we don't signal an event saying the model is reloaded until all of its dependencies are ready. So I think it will add some more complexity, and will not be relevant long term since we want to squash this hierarchy down eventually anyways.  So I'd just as soon keep it as-is for now.

Reply:
So you can't update the nested assets because they are all encapsulated.  If assets are ever going to be this complex they should have an official, friend, asset (re)loader class similar to what is done with creators etc.